enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -5001,6 +5000,38 @@ class AdjointGenerator
             }
           } else
             assert(0 && "illegal mpi");
+        } else if (Mode == DerivativeMode::ForwardMode) {
+          IRBuilder<> Builder2(&call);
+          getForwardBuilder(Builder2);
+
+          assert(!gutils->isConstantValue(call.getOperand(0)));
+          assert(!gutils->isConstantValue(call.getOperand(6)));
+
+          Value *buf = gutils->invertPointerM(call.getOperand(0), Builder2);
+          Value *count = gutils->getNewFromOriginal(call.getOperand(1));
+          Value *datatype = gutils->getNewFromOriginal(call.getOperand(2));
+          Value *source = gutils->getNewFromOriginal(call.getOperand(3));
+          Value *tag = gutils->getNewFromOriginal(call.getOperand(4));
+          Value *comm = gutils->getNewFromOriginal(call.getOperand(5));
+          Value *request = gutils->invertPointerM(call.getOperand(6), Builder2);
+
+          Value *args[] = {
+              /*buf*/ buf,
+              /*count*/ count,
+              /*datatype*/ datatype,
+              /*source*/ source,
+              /*tag*/ tag,
+              /*comm*/ comm,
+              /*request*/ request,
+          };
+
+          Type *types[sizeof(args) / sizeof(*args)];
+          for (size_t i = 0; i < sizeof(args) / sizeof(*args); i++)
+            types[i] = args[i]->getType();
+
+          FunctionType *FT = FunctionType::get(call.getType(), types, false);
```

Comment:
In this case can you use getCalledValue (llvm version appropriate). And use that to get the corresponding function type?

Reply:
fixed